KIPs/kip-113.md

```diff
@@ -0,0 +1,252 @@
+| ---------------------------- | -------------- | ------------------------- | ------------------------------ | --------------- | --------------------------------- | ---------------------- |
+| Binary modification required | O              | O                         | X<br>Patch AddressBook address | O               | X<br> Patch the hardcoded address | O                      |
+| Contract deployment required | X<br> Registry | X <br> Registry, GovParam | X <br> Registry, AddressBook   | X<br> Registry  | X<br> Registry                    | X<br> AddressBook      |
+
+In terms of usability from ServiceChain, HeaderGov, ContractGov, ChainConfig, and Bytecode-injection are recommended.
+
+##### Others
+
+| Methods                                          | [1] HeaderGov                 | [2] ContractGov               | [3] SpareContract             | [4] ChainConfig      | [5] Precompiled-contract | [6] Bytecode-injection |
+| ------------------------------------------------ | ----------------------------- | ----------------------------- | ----------------------------- | -------------------- | ------------------------ | ---------------------- |
+| Possible to panic outdated nodes at hardfork     | X<br> At GovernanceData block | X<br> At the activation block | X<br> At SpareContract update | O                    | O                        | O                      |
+| Possible to setup initial data before hardfork   | O                             | O                             | O                             | O                    | O                        | X                      |
+| Possible to use a special address (like 0x0…400) | X<br> vanity address          | X<br> vanity address          | X<br> vanity address          | X<br> vanity address | O                        | O                      |
+
+In terms of various remaining aspects, ChainConfig, Precompiled-contract, and Bytecode-injection are recommended.
+
+Overall, we choose ChainConfig because:
+
+- The main subscriber is not smart contracts but rather the Core protocol, so the awareness of subscriber contracts after address update is less important.
+- Ease of use in all of ServiceChain, Baobab (testnet), Cypress (mainnet).
+- Initial data setup before hardfork is viable.
+- Special address format is less important.
+
+## Backwards Compatibility
+
+This does not affect the backward compatibility as this is a newly deployed contract.
+
+## Implementation
+
+A reference implementation: [link](https://github.com/klaytn/simple-bls-registry)
+
+### Cypress-specific implementation
+
+Cypress has the following access controls:
+
+- `registerPublicKey` reverts if `msg.sender` is not a `cnNodeId` registered in AddressBook.
+- `unregisterPublicKey(cnNodeId)` reverts if `msg.sender` is not an admin of AddressBook, or if `cnNodeId` is in AddressBook.
```

Comment:
```suggestion
- `unregisterPublicKey(cnNodeId)` reverts if `msg.sender` is not an admin of AddressBook, or if `cnNodeId` is not in AddressBook.
```

Reply:
This section is removed because a representative registry will not be deployed in Cypress as a system contract.